File: java/src/com/sm/Coordinator.java
package com.sm;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.UUID;

public class Coordinator {

    int PORT = 4999;
    boolean listeningClients = true;
    boolean acceptingConnections = true;
    ServerSocket serverSocket;
    ArrayList<Socket> clientSockets;
    ArrayList<Socket> searchingClients;
    JSONArray cachedMatches2;
    ArrayList<JSONObject> serverList = new ArrayList<>(Arrays.asList(
            new JSONObject().put("address", "127.0.0.1").put("port", 5000),
            new JSONObject().put("address", "127.0.0.1").put("port", 5001),
            new JSONObject().put("address", "127.0.0.1").put("port", 5002),
            new JSONObject().put("address", "127.0.0.1").put("port", 5003),
            new JSONObject().put("address", "127.0.0.1").put("port", 5004)
    ));

    public Coordinator() throws IOException {
        serverSocket = new ServerSocket(PORT);
        clientSockets = new ArrayList<>();
        searchingClients = new ArrayList<>();
        cachedMatches2 = new JSONArray();
        checkStatus();
        handleConnections();
        listenConnections();
    }

    InetSocketAddress checkStatus() {
        ArrayList<JSONObject> serverInformation = new ArrayList<>();
        InetSocketAddress selectedServer = null;
        try {
            for (JSONObject server : serverList) {
                try {
                    byte[] receiveBuffer = new byte[150];
                    byte[] sendBuffer = "{\"data-type\": \"query\"}".getBytes();
                    int serverPort = server.getInt("port");
                    String serverAddress = server.getString("address");
                    DatagramSocket querySocket = new DatagramSocket();
                    DatagramPacket sendPacket = new DatagramPacket(sendBuffer, sendBuffer.length, InetAddress.getByName(serverAddress), serverPort);
                    DatagramPacket receivePacket = new DatagramPacket(receiveBuffer, receiveBuffer.length);
                    querySocket.setSoTimeout(1000);
                    querySocket.send(sendPacket);
                    querySocket.receive(receivePacket);
                    querySocket.setSoTimeout(0);
                    String receivedData = new String(receivePacket.getData(), 0, receivePacket.getLength());
                    JSONObject information = new JSONObject(receivedData);
                    if (information.get("data-type").equals("response")) {
                        information.put("address", serverAddress).put("port", serverPort);
                        serverInformation.add(information);
                        // updateStatus(information);
                        if (information.get("phase").equals("waiting")) {
                            System.out.printf("('%s', %d) is available\n", serverAddress, serverPort); // format same as other versions
                            selectedServer = new InetSocketAddress(InetAddress.getByName(serverAddress), serverPort);
                        } else {
                            System.out.printf("('%s', %d) is busy\n", serverAddress, serverPort);
                        }
                    }
                } catch (SocketTimeoutException ir) {
                    System.out.printf("('%s', %d) is unresponsive\n", server.getString("address"), server.getInt("port"));
                }
            }
        } catch (Exception e) { e.printStackTrace(); }
        for (JSONObject i: serverInformation) {
            updateStatus(i);
        }
        return selectedServer;
    }

    void updateStatus(JSONObject serverInformation) {
        int indexCounter = 0;
        JSONArray cachedMatches3 = cachedMatches2;
        for (Object i: cachedMatches2) {
            JSONObject information = (JSONObject) i;
            if (information.getString("address").equals(serverInformation.getString("address"))) {
                if (information.getInt("port") == serverInformation.getInt("port")) {
                    cachedMatches3.remove(indexCounter);
                }
            }
            indexCounter++;
        }
        cachedMatches3.put(serverInformation);
        cachedMatches2 = cachedMatches3;
    }

    public void sendMessage(String message, Socket client) {
        try {
            PrintWriter outputStream = new PrintWriter(client.getOutputStream(), true);
            outputStream.println(message);
            outputStream.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void prepareServer(Socket clientOne, Socket clientTwo, InetSocketAddress serverAddress) {
        try {
            String[] connectionTokens = {UUID.randomUUID().toString(), UUID.randomUUID().toString()};
            JSONObject prepareRequest = new JSONObject();
            JSONObject tokenRequest = new JSONObject();
            DatagramSocket serverSocket = new DatagramSocket();
            DatagramPacket sendPacket;
            prepareRequest.put("data-type", "prepare");
            prepareRequest.put("token_1", connectionTokens[0]);
            prepareRequest.put("token_2", connectionTokens[1]);
            tokenRequest.put("data-type", "match");
            tokenRequest.put("address", serverAddress.getAddress().getHostAddress());
            tokenRequest.put("port", serverAddress.getPort());
            tokenRequest.put("token", connectionTokens[0]);
            sendPacket = new DatagramPacket(prepareRequest.toString().getBytes(), 0, prepareRequest.toString().getBytes().length, serverAddress);
            serverSocket.send(sendPacket);
            sendMessage(tokenRequest.toString(), clientOne);
            tokenRequest.put("token", connectionTokens[1]);
            sendMessage(tokenRequest.toString(), clientTwo);

        } catch (IOException e) { e.printStackTrace(); }
    }

    void matchmakeClients() {
        if (searchingClients.size() >= 2) {
            ArrayList<Socket> copiedSearchers = new ArrayList<>(searchingClients);
            InetSocketAddress selectedServer = checkStatus();
            if (selectedServer != null) {
                prepareServer(copiedSearchers.get(0), copiedSearchers.get(1), selectedServer);
                searchingClients.remove(copiedSearchers.get(0));
                searchingClients.remove(copiedSearchers.get(1));
            }
        }
    }

    void _listenConnections() throws IOException {
        while (listeningClients) {
            ArrayList<Socket> copiedClients = new ArrayList<>(clientSockets.size());
            copiedClients.addAll(clientSockets);
            for (Socket clientSocket: copiedClients) {
                BufferedReader inputStream = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
                String receivedData;
                try {
                    if (inputStream.ready()) {
                        if ((receivedData = inputStream.readLine()) != null) {
                            System.out.println(receivedData);
                            JSONObject receivedInformation = new JSONObject(receivedData);
                            if (receivedInformation.getString("data-type").equalsIgnoreCase("search")) {
                                searchingClients.add(clientSocket);
                                matchmakeClients();
                            }
                        }
                    }
                } catch (Exception e) {
                    clientSockets.remove(clientSocket);
                    System.out.println("Client " + clientSocket.getRemoteSocketAddress() + " has disconnected from the server.");
                }
            }
        }
    }

    void listenConnections() {
        new Thread(() -> {
            try {
                _listenConnections();
            } catch (IOException e) { e.printStackTrace(); }
        }).start();
    }

    void _handleConnections() throws IOException {
        while (acceptingConnections) {
            Socket clientSocket = serverSocket.accept();
            clientSockets.add(clientSocket);
            System.out.println("Client " + clientSocket.getRemoteSocketAddress() + " has connected to the server.");
        }
    }

    void handleConnections() {
        new Thread(() -> {
            try {
                _handleConnections();
            } catch (IOException e) { e.printStackTrace(); }
        }).start();
    }

}

File: python/coordinator.py
from socket import socket, AF_INET, SOCK_STREAM, SOCK_DGRAM
from select import select
from ast import literal_eval
from threading import Thread
from time import sleep
from secrets import token_hex


class Pong:
    def __init__(self):
        self.socket = socket(AF_INET, SOCK_STREAM)
        self.socket.bind(("0.0.0.0", 4999))
        self.socket.listen(10)
        self.clients = [self.socket]
        self.cached_matches = {}
        self.searching_clients = []
        self.server_list = [
            ("127.0.0.1", 5000),
            ("127.0.0.1", 5001),
            ("127.0.0.1", 5002),
            ("127.0.0.1", 5003),
            ("127.0.0.1", 5004)
            ]
        Thread(target=self.check_status).start()
        self.listen()

    def _check_status(self):
        print("--")
        for server in self.server_list:
            try:
                query_socket = socket(AF_INET, SOCK_DGRAM)
                query_socket.connect(server)
                query_socket.settimeout(10)
                query_socket.send(str.encode(str({"data-type": "query"})))
                response, address = query_socket.recvfrom(150)
                information = literal_eval(response.decode())
                if information["data-type"] == "response":
                    self.cached_matches[server] = information
                    if information["phase"] == "waiting":
                        print(str(server) + " is available")
                    else:
                        print(str(server) + " is busy")
            except (TimeoutError, ConnectionResetError, Exception):
                print(str(server) + " is unresponsive")

    def check_status(self):
        while True:
            self._check_status()
            self.matchmake()
            sleep(10)

    def server_available(self):
        for server in self.server_list:
            try:
                query_socket = socket(AF_INET, SOCK_DGRAM)
                query_socket.connect(server)
                query_socket.settimeout(10)
                query_socket.send(str.encode(str({"data-type": "query"})))
                response, address = query_socket.recvfrom(150)
                information = literal_eval(response.decode())
                if information["data-type"] == "response":
                    if information["phase"] == "waiting":
                        return server
            except (TimeoutError, ConnectionResetError, Exception):
                print(str(server) + " did not respond to the query.")
        return False

    def prepare(self, client1, client2, server):
        print("prep")
        tokens = [token_hex(16), token_hex(16)]
        server_socket = socket(AF_INET, SOCK_DGRAM)
        server_socket.connect(server)
        server_socket.settimeout(10)
        server_socket.send(str.encode(str({"data-type": "prepare", "token_1": tokens[0], "token_2": tokens[1]})))
        #sleep(1)
        client1.send(str.encode(str({"data-type": "match", "address": server[0], "port": server[1], "token": tokens[0]}) + '\n'))
        client2.send(str.encode(str({"data-type": "match", "address": server[0], "port": server[1], "token": tokens[1]}) + '\n'))

    def matchmake(self):
        if len(self.searching_clients) >= 2:
            copied_searchers = self.searching_clients.copy()
            if server := self.server_available():
                self.prepare(copied_searchers[0], copied_searchers[1], server)
                self.searching_clients.remove(copied_searchers[0])
                self.searching_clients.remove(copied_searchers[1])
                print("Client [%s:%d] joined a match." % (
                copied_searchers[0].getpeername()[0], copied_searchers[0].getpeername()[1]))
                print("Client [%s:%d] joined a match." % (
                copied_searchers[1].getpeername()[0], copied_searchers[1].getpeername()[1]))

    def listen(self):
        while True:
            try:
                read_sockets, write_sockets, error_sockets = select(self.clients, [], [])
                for sock in read_sockets:
                    if sock == self.socket:
                        sockfd, address = self.socket.accept()
                        self.clients.append(sockfd)
                        print(f'Client [{address[0]}:{address[1]}] connected to the server.')
                    else:
                        try:
                            received_data = sock.recv(150).decode()
                        except:
                            print("Client [%s:%d] disconnected from the server." % (sock.getpeername()[0], sock.getpeername()[1]))
                            sock.close()
                            self.clients.remove(sock)
                            continue
                        if not received_data:
                            continue
                        information = literal_eval(received_data)
                        if information["data-type"] == "search":
                            self.searching_clients.append(sock)
                            self.matchmake()
            except Exception as error:
                print(error)




if __name__ == '__main__':
    Pong()
